RFC-0031-ZenDNN-Integration.md

```diff
@@ -0,0 +1,94 @@
+
+# ZenDNN Integration
+
+
+## **Author**
+* @naveenthangudu
+
+
+## **Summary**
+This document proposes an approach for integrating ZenDNN library into PyTorch. This integration will enable inference optimizations for deep learning workloads on AMD CPUs.
+
+
+## **Highlights**
+* Default build behavior remains un-altered.
+* A pre-build step will generate ZenDNN integration code.
+* ZenDNN integration is optimized for inference workloads.
+* New opt-in build flag (environmental variable) `USE_ZENDNN` will be added.
+* No Submodules are added. When `USE_ZENDNN` is set to 1, the following repositories will be downloaded into third_party folder.
+  * [ZenDNN](https://github.com/amd/ZenDNN)
+  * [AOCL BLIS](https://github.com/amd/blis)
+* ZenDNN optimizations are added as JIT Passes on frozen graph under `optimize_for_inference` API.
+* Support will be added for fp32 and bf16 data types.
+
+
+## **Motivation**
+The main goal of this integration is to realize optimal inference on AMD CPUs for PyTorch. ZenDNN is a library which enables performance improvements on AMD CPU architectures. Please find the repo and more details [here](https://github.com/amd/ZenDNN). To highlight the potential uplift using ZenDNN vs. oneDNN on AMD CPUs, we have benchmarked our fully integrated ZenDNN version using PyTorch v1.12 on a Genoa CPU (4th Generation AMD EPYC™ Processors).
+
+
+### **Benchmarking Configuration**
+|||
+|-----------------------------|-------|
+| PyTorch version             | v1.12  |
+| CPU                         | 4th Generation AMD EPYC™ Genoa CPU |
+| Number of physical cores    | 96   |
+| NUMA Nodes per Socket(NPS)  | 1    |
+| Model data type             | FP32 |
+
+### **Latency Performance - batch_size=1**
+
+<img src="./RFC-0031-assets/Latency.svg" width=40%>
+
+### **Throughput Performance - batch_size=640**
+
+<img src="./RFC-0031-assets/Throughput.svg" width=40%>
+
+
+## **Proposed Implementation**
+
+### **Overview**
+This approach references the [hipify](https://github.com/ROCm-Developer-Tools/HIPIFY) approach used by the AMD ROCm team for integration of ROCm libraries into PyTorch. By leveraging shared APIs between oneDNN and ZenDNN, a pre-build step will generate ZenDNN integration code wherever possible. By adding an opt-in build flag, default build behavior is un-altered. Required repositories are only downloaded when the build flag is enabled.
```

Comment:
oneDNN comes with both primitive API (https://oneapi-src.github.io/oneDNN/supported_primitives.html) and graph API (https://oneapi-src.github.io/oneDNN/graph_extension.html). Are you going to replace both?

What oneDNN version you are based on? We will upgrade to oneDNN 3.x in PyTorch soon. New features and optimizations will be available for both x86 CPU and ARM CPU. How are you going to maintain the API compatibility with oneDNN in the future?

Can you explore options other than the "hipify" method and compare their advantages and disadvantages with "hipify"? For instance, one possibility is to incorporate your optimization improvements directly into oneDNN and use the pre-existing integration with PyTorch, which is the path that ARM has taken.

Reply:
> oneDNN comes with both primitive API (https://oneapi-src.github.io/oneDNN/supported_primitives.html) and graph API (https://oneapi-src.github.io/oneDNN/graph_extension.html). Are you going to replace both?

We are planning to upstream changes relevant to oneDNN primitive API.

> What oneDNN version you are based on?

Currently ZenDNN leverages oneDNN v2.6.3's basic infrastructure and APIs. Please check README.md in ZenDNN [repo](https://github.com/amd/ZenDNN).

> We will upgrade to oneDNN 3.x in PyTorch soon. New features and optimizations will be available for both x86 CPU and ARM CPU. How are you going to maintain the API compatibility with oneDNN in the future?

As pytorch is integrated with ideep instead of oneDNN. Are you planning to upgrade the ideep API also?

> Can you explore options other than the "hipify" method and compare their advantages and disadvantages with "hipify"? For instance, one possibility is to incorporate your optimization improvements directly into oneDNN and use the pre-existing integration with PyTorch, which is the path that ARM has taken.

We are following hipify approach based on recommendation from this [thread](https://github.com/pytorch/pytorch/pull/76242).

Reply:
> As pytorch is integrated with ideep instead of oneDNN. Are you planning to upgrade the ideep API also?

Yes. We will upgrade ideep to support oneDNN 3.x shortly. There is also request from community to upgrade oneDNN 3.x too. https://github.com/pytorch/pytorch/issues/97757

Reply:
Is the ideep API/interface going to change when the oneDNN is upgraded from 2.x to 3.x ?

Reply:
My dream is that some day we would retire ideep for good...